tests/integration_tests/functional/test_cpu_features.py

```diff
@@ -314,6 +346,367 @@ def _test_cpu_rdmsr(context):
+def dump_msr_state_to_file(dump_fname, ssh_conn, shared_names):
+    """
+    Read MSR state via SSH and dump it into a file.
+    """
+    ssh_conn.scp_file(
+        shared_names["msr_reader_host_fname"], shared_names["msr_reader_guest_fname"]
+    )
+    _, stdout, stderr = ssh_conn.execute_command(shared_names["msr_reader_guest_fname"])
+    assert stderr.read() == ""
+
+    with open(dump_fname, "w", encoding="UTF-8") as file:
+        file.write(stdout.read())
+
+
+def _test_cpu_wrmsr_snapshot(context):
+    shared_names = context.custom["shared_names"]
+    root_disk = context.disk.copy(file_name=shared_names["rootfs_fname"])
+    vm_builder = context.custom["builder"]
+    cpu_template = context.custom["cpu_template"]
+
+    vm_instance = vm_builder.build(
+        kernel=context.kernel,
+        disks=[root_disk],
+        ssh_key=context.disk.ssh_key(),
+        config=context.microvm,
+        diff_snapshots=True,
+        cpu_template=cpu_template,
+    )
+
+    vm = vm_instance.vm
+    vm.start()
+
+    # Make MSR modifications
+    ssh_connection = net_tools.SSHConnection(vm.ssh_config)
+
+    msr_writer_host_fname = "../resources/tests/msr/msr_writer.sh"
+    msr_writer_guest_fname = "/bin/msr_writer.sh"
+    ssh_connection.scp_file(msr_writer_host_fname, msr_writer_guest_fname)
+
+    wrmsr_input_host_fname = "../resources/tests/msr/wrmsr_list.txt"
+    wrmsr_input_guest_fname = "/tmp/wrmsr_input.txt"
+    ssh_connection.scp_file(wrmsr_input_host_fname, wrmsr_input_guest_fname)
+
+    _, _, stderr = ssh_connection.execute_command(
+        f"{msr_writer_guest_fname} {wrmsr_input_guest_fname}"
+    )
+    assert stderr.read() == ""
+
+    # Dump MSR state to a file that will be published to S3 for the 2nd part of the test
+    snapshot_artifacts_dir = os.path.join(
+        shared_names["snapshot_artifacts_root_dir_wrmsr"],
+        context.kernel.base_name(),
+        cpu_template if cpu_template else "none",
+    )
+    os.makedirs(snapshot_artifacts_dir)
+
+    msrs_before_fname = os.path.join(
+        snapshot_artifacts_dir, shared_names["msrs_before_fname"]
+    )
+
+    dump_msr_state_to_file(msrs_before_fname, ssh_connection, shared_names)
+
+    # Take a snapshot
+    vm.pause_to_snapshot(
+        mem_file_path=shared_names["mem_fname"],
+        snapshot_path=shared_names["snapshot_fname"],
+        diff=True,
+    )
+
+    # Copy snapshot files to be published to S3 for the 2nd part of the test
+    chroot_dir = vm.chroot()
+    shutil.copyfile(
+        os.path.join(chroot_dir, shared_names["mem_fname"]),
+        os.path.join(snapshot_artifacts_dir, shared_names["mem_fname"]),
+    )
+    shutil.copyfile(
+        os.path.join(chroot_dir, shared_names["snapshot_fname"]),
+        os.path.join(snapshot_artifacts_dir, shared_names["snapshot_fname"]),
+    )
+    shutil.copyfile(
+        root_disk.local_path(),
+        os.path.join(snapshot_artifacts_dir, shared_names["rootfs_fname"]),
+    )
+
+
+@pytest.mark.skipif(
```

Comment:
are all tests in this file x86 specific? We can skip the whole file at the top rather than one by one https://docs.pytest.org/en/7.1.x/reference/reference.html#globalvar-pytestmark

```
pytestmark = pytest.mark.skipif(...)
```

Reply:
there is one old test that runs on Arm... I guess we could look into restructuring of this file when user-configurable templates are available and new tests are implemented.